config/examples/transparent.yml

```diff
@@ -0,0 +1,19 @@
+log_debug: true
```

Comment:
Until a configuration does not use cache, allow_transparent_cache is not needed.
I assume that while cache is not configured, it is disabled. Is it wrong?

Reply:
oh I misunderstood. What's the advantage of introducing `allow_transparent_cache` then? If user wants to add caching he will need to link it in the config section as usual, if not otherwise.

Reply:
The idea of transparent_user is to represent several real CH users.
These users would share cache with each other.
I consider this as a potential security risk e.g. because of Row Level Security malfunctioning.
That is why I've decided to make this risk more obvious, _allow_transparent_cache_ is somewhat similar to _hack_me_please._

Reply:
What do you think about adding a config parameter to the user section explicitly specifying if it's an ldap clickhouse user? This way no such mechanism would be needed (I think). It's a bit unintuitive to me

Reply:
Not definitely LDAP, maybe just Kerberos. Another possible concern - plain passwords in chproxy config someone would like to avoid.
So, any user we would like to auth using ClickHouse, not chproxy facilities.

Using a magic name does not seem perfect, agree, but I do not fully understand your idea.

If you suggest something like
```
users:
  - name: "some_user"
    to_cluster: "default_with_transparent"
    to_user: "some_user"
    transparency: true
clusters:
  - name: "default_with_transparent"
    nodes: ["127.0.0.1:8123"]

    users:
    - name: "some_user"
      transparency: true
```
it is doable of course, we do not use a magic name anymore, but use an explicit flag. 
Is it better from your point of view?

The whole idea is to be able to pass original credentials to ClickHouse for some (not all) users.
It is impossible to have the real usernames of such users in chproxy config.
That is why we need two types of users in chproxy configuration anyway.

Reply:
did you think of creating an LDAP connector in order to fetch the users instead of creating this transparent_user concept that breaks a bit the concept of users in CHProxy (cf the issues with caching)?

Reply:
Hello Christophe, 
I am not sure that I fully understand the idea: if a user is not listed in chproxy config, how his or her properties (e.g. quotas) can be defined? It does not look much better than transparent_user, but definitely more complex.

I agree with you, transparent user concept is not in line with the concept of chproxy users. 
But we should face reality, native ClickHouse authentication facilities are rather advanced.
They include
- LDAP support. There are several options, e.g Role Based Access Control.
- Kerberos support (which is a special problem).
- native user management via SQL with passwords in sha256
So, it is not just about LDAP. 
Speaking of which, setting up LDAP is rather cumbersome sometimes, and the idea of doing it twice is not necessarily attractive for Operations.

What can be done to make chproxy users great again? 
For example, we can introduce wildcards. 
Instead of a single magic transparent_user we can have users like *_dba , *_departmet1_analyst, *_department2_analyst. These transparent (or umbrella or whatever called) users can be maintained by chproxy from quotas and cache point of view, leaving real authenticating and authorization for ClickHouse. 
This approach seems to be safe and convenient.
@mga-chka , what do you think?

Reply:
Prefixes might be better than suffixes, dba_*, not *_dba

Reply:
Sorry for not answering now. It just to let you know that we're discussing with pawel about your PR (because it's conceptually impactful) but since we're busy on other tasks we might give an answer at the beginning of next week.

Reply:
Hi, 

We believe that having a specific LDAP support inside CHProxy is the cleaner way regarding the current logic of CH proxy (even if it means duplicating the logic of CH and doing the same for kerberos and so on).
That being said, you're right that CHProxy is far from what CH provides in terms of user management.

So, here is our proposal: having a specific config flag that will deactivate ALL the user logic and the features depending on users (caching, quotas). In this mode, CHProxy would be just a simple proxy whose job is to reach on off the clusters nodes.
Letting people use users based features without the user concepts will for sure lead to bugs so this is why we don't want that for the moment.

Nb: if we rework the user management, nothing prevent us in the future to activate cleanly the user based features even if the users are managed only by CH

WDYT?

Reply:
Hello Christophe,
I've added a description of wildcarded users concept to the PR https://hackmd.io/LXk8tLI8Q3etKax-HWtGxQ .
To me, it is not that bad, actually, it has some advantages.
After all, chproxy customers can decide if they take risks of 'impersonation' or not.
On the other hand, your concept looks solid, it is unbeatable. Maybe it is just too perfect for our non-perfect world? ;)
What about having the flag that disables user machinery completely AND having wildcarded users?

Reply:
Hi, I left the following comments a few days ago in your documents 
"[christophe] what about the caching part? Do we assume that like for quota it will be shared by all the department?"
I'm assuming the answer will be yes. If so then ok, let's do your proposal.

Reply:
since you need to rewrite a part of your PR, can you please avoid the use of "**" (for example in getUser() or getTransparentUser() functions)?
I know we're picky but we want to keep the code consistent and no one has used a pointer of pointer in chproxy codebase until now.
nb: We will soon add code guidelines so that it will be more explicit

Reply:
Hello Christophe,
thank you for your response.
Yes, I assume the same, the cache "will be shared by all the department".
Cache works for a wildcarded_user/umbrella _user/department/whatever_it_called in the same way as for normal user.

So we are going with wilcarded user concept. 
Cool!
Should we create another PR or it is Ok to continue in current PR and current branch?

Reply:
If you can, create a new PR because the code will change a lot and this PR already has a lot of comments.

Reply:
Hello Christophe @mga-chka , sorry for the delay. 
Vacation is over and I am back to work.
Could you help me with the name of these new users, please? Is 'wildcarded user' Ok?

Reply:
Hello @mga-chka , could you help me with the name of these new users, please?
Going with "wildcarded user" so far.

Reply:
my bad, I did see this notification then I was off. Yes let's go for this name (since I don't have something better)